File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.java
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2012 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc;

import android.app.Dialog;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.Window;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;
import android.support.v4.app.ActionBarDrawerToggle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.AppCompatActivity;

import edu.berkeley.boinc.adapter.NavDrawerListAdapter;
import edu.berkeley.boinc.adapter.NavDrawerListAdapter.NavDrawerItem;
import edu.berkeley.boinc.attach.SelectionListActivity;
import edu.berkeley.boinc.client.ClientStatus;
import edu.berkeley.boinc.client.Monitor;
import edu.berkeley.boinc.client.IMonitor;
import edu.berkeley.boinc.rpc.Project;
import edu.berkeley.boinc.utils.BOINCDefs;
import edu.berkeley.boinc.utils.Logging;

import java.util.ArrayList;

public class BOINCActivity extends AppCompatActivity {

    public static IMonitor monitor;
    private Integer clientComputingStatus = -1;
    private Integer numberProjectsInNavList = 0;
    static Boolean mIsBound = false;

    // app title (changes with nav bar selection)
    private CharSequence mTitle;
    // nav drawer title
    private CharSequence mDrawerTitle;

    private DrawerLayout mDrawerLayout;
    private ListView mDrawerList;
    private ActionBarDrawerToggle mDrawerToggle;
    private NavDrawerListAdapter mDrawerListAdapter;

    private ServiceConnection mConnection = new ServiceConnection() {
        public void onServiceConnected(ComponentName className, IBinder service) {
            // This is called when the connection with the service has been established, getService returns
            // the Monitor object that is needed to call functions.
            monitor = IMonitor.Stub.asInterface(service);
            mIsBound = true;
            determineStatus();
        }

        public void onServiceDisconnected(ComponentName className) {
            // This should not happen
            monitor = null;
            mIsBound = false;

            Log.e(Logging.TAG, "BOINCActivity onServiceDisconnected");
        }
    };

    private BroadcastReceiver mClientStatusChangeRec = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if(Logging.VERBOSE) {
                Log.d(Logging.TAG, "BOINCActivity ClientStatusChange - onReceive()");
            }
            determineStatus();
        }
    };
    private IntentFilter ifcsc = new IntentFilter("edu.berkeley.boinc.clientstatuschange");

    @Override
    public void onCreate(Bundle savedInstanceState) {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onCreate()");
        }
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);

        // setup navigation bar
        mTitle = mDrawerTitle = getTitle();
        mDrawerLayout = findViewById(R.id.drawer_layout);
        mDrawerList = findViewById(R.id.list_slidermenu);
        mDrawerList.setOnItemClickListener(new ListView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                // display view for selected nav drawer item
                dispatchNavBarOnClick(mDrawerListAdapter.getItem(position), false);
            }
        });
        mDrawerListAdapter = new NavDrawerListAdapter(getApplicationContext());
        mDrawerList.setAdapter(mDrawerListAdapter);
        // enabling action bar app icon and behaving it as toggle button
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_drawer);

        mDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayout,
                                                  R.drawable.ic_drawer, //nav menu toggle icon
                                                  R.string.app_name, // nav drawer open - description for accessibility
                                                  R.string.app_name // nav drawer close - description for accessibility
        ) {
            public void onDrawerClosed(View view) {
                getSupportActionBar().setTitle(mTitle);
                // calling onPrepareOptionsMenu() to show action bar icons
                supportInvalidateOptionsMenu();
            }

            public void onDrawerOpened(View drawerView) {
                getSupportActionBar().setTitle(mDrawerTitle);
                mDrawerListAdapter.notifyDataSetChanged(); // force redraw of all items (adapter.getView()) in order to adapt changing icons or number of tasks/notices
                // calling onPrepareOptionsMenu() to hide action bar icons
                supportInvalidateOptionsMenu();
            }
        };
        mDrawerLayout.setDrawerListener(mDrawerToggle);


        // pre-select fragment
        // 1. check if explicitly requested fragment present
        // e.g. after initial project attach.
        int targetFragId = getIntent().getIntExtra("targetFragment", -1);

        // 2. if no explicit request, try to restore previous selection
        if(targetFragId < 0 && savedInstanceState != null) {
            targetFragId = savedInstanceState.getInt("navBarSelectionId");
        }

        NavDrawerItem item;
        if(targetFragId < 0) {
            // if non of the above, go to default
            item = mDrawerListAdapter.getItem(0);
        }
        else {
            item = mDrawerListAdapter.getItemForId(targetFragId);
        }

        if(item != null) {
            dispatchNavBarOnClick(item, true);
        }
        else if(Logging.WARNING) {
            Log.w(Logging.TAG, "onCreate: fragment selection returned null");
        }

        //bind monitor service
        doBindService();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putInt("navBarSelectionId", mDrawerListAdapter.selectedMenuId);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onDestroy() {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onDestroy()");
        }
        doUnbindService();
        super.onDestroy();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onNewIntent()");
        }
        // onNewIntent gets called if activity is brought to front via intent, but was still alive, so onCreate is not called again
        // getIntent always returns the intent activity was created of, so this method is the only hook to receive an updated intent
        // e.g. after (not initial) project attach
        super.onNewIntent(intent);
        // navigate to explicitly requested fragment (e.g. after project attach)
        int id = intent.getIntExtra("targetFragment", -1);
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onNewIntent() for target fragment: " + id);
        }
        NavDrawerItem item = mDrawerListAdapter.getItemForId(id);
        if(item != null) {
            dispatchNavBarOnClick(item, false);
        }
        else if(Logging.WARNING) {
            Log.w(Logging.TAG, "onNewIntent: requested target fragment is null, for id: " + id);
        }
    }

    @Override
    protected void onResume() { // gets called by system every time activity comes to front. after onCreate upon first creation
        super.onResume();
        registerReceiver(mClientStatusChangeRec, ifcsc);
        determineStatus();
    }

    @Override
    protected void onPause() { // gets called by system every time activity loses focus.
        if(Logging.VERBOSE) {
            Log.v(Logging.TAG, "BOINCActivity onPause()");
        }
        super.onPause();
        unregisterReceiver(mClientStatusChangeRec);
    }

    private void doBindService() {
        // start service to allow setForeground later on...
        startService(new Intent(this, Monitor.class));
        // Establish a connection with the service, onServiceConnected gets called when
        bindService(new Intent(this, Monitor.class), mConnection, Service.BIND_AUTO_CREATE);
    }

    private void doUnbindService() {
        if(mIsBound) {
            // Detach existing connection.
            unbindService(mConnection);
            mIsBound = false;
        }
    }
	/*
	public IMonitor getMonitorService() {
		if(!mIsBound) if(Logging.WARNING) Log.w(Logging.TAG, "Fragment trying to obtain serive reference, but Monitor not bound in BOINCActivity");
		return monitor;
	}*/

    public void startAttachProjectListActivity() {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity attempt to start ");
        }
        startActivity(new Intent(this, SelectionListActivity.class));
    }

    /**
     * React to selection of nav bar item
     *
     * @param item Nav bar item
     * @param init Initialize
     */
    private void dispatchNavBarOnClick(NavDrawerItem item, boolean init) {
        // update the main content by replacing fragments
        if(item == null) {
            if(Logging.WARNING) {
                Log.w(Logging.TAG, "dispatchNavBarOnClick returns, item null.");
            }
            return;
        }
        if(Logging.DEBUG) {
            Log.d(Logging.TAG,
                  "dispatchNavBarOnClick for item with id: " + item.getId() + " title: " + item.getTitle() +
                  " is project? " + item.isProjectItem());
        }

        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        Boolean fragmentChanges = false;
        if(init) {
            // if init, setup status fragment
            ft.replace(R.id.status_container, new StatusFragment());
        }
        if(!item.isProjectItem()) {
            switch(item.getId()) {
                case R.string.tab_tasks:
                    ft.replace(R.id.frame_container, new TasksFragment());
                    fragmentChanges = true;
                    break;
                case R.string.tab_notices:
                    ft.replace(R.id.frame_container, new NoticesFragment());
                    fragmentChanges = true;
                    break;
                case R.string.tab_projects:
                    ft.replace(R.id.frame_container, new ProjectsFragment());
                    fragmentChanges = true;
                    break;
                case R.string.menu_help:
                    startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://boinc.berkeley.edu/wiki/BOINC_Help")));
                    break;
                case R.string.menu_report_issue:
                    startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://boinc.berkeley.edu/trac/wiki/ReportBugs")));
                    break;
                case R.string.menu_about:
                    final Dialog dialog = new Dialog(this);
                    dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
                    dialog.setContentView(R.layout.dialog_about);
                    Button returnB = dialog.findViewById(R.id.returnB);
                    TextView tvVersion = dialog.findViewById(R.id.version);
                    try {
                        tvVersion.setText(getString(R.string.about_version) + " "
                                          + getPackageManager().getPackageInfo(getPackageName(), 0).versionName);
                    }
                    catch(NameNotFoundException e) {
                        if(Logging.WARNING) {
                            Log.w(Logging.TAG, "version name not found.");
                        }
                    }

                    returnB.setOnClickListener(new OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            dialog.dismiss();
                        }
                    });
                    dialog.show();
                    break;
                case R.string.menu_eventlog:
                    startActivity(new Intent(this, EventLogActivity.class));
                    break;
                case R.string.projects_add:
                    startActivity(new Intent(this, SelectionListActivity.class));
                    break;
                case R.string.tab_preferences:
                    ft.replace(R.id.frame_container, new PrefsFragment());
                    fragmentChanges = true;
                    break;

                default:
                    if(Logging.ERROR) {
                        Log.d(Logging.TAG,
                              "dispatchNavBarOnClick() could not find corresponding fragment for " + item.getTitle());
                    }
                    break;
            }

        }
        else {
            // ProjectDetailsFragment. Data shown based on given master URL
            Bundle args = new Bundle();
            args.putString("url", item.getProjectMasterUrl());
            Fragment frag = new ProjectDetailsFragment();
            frag.setArguments(args);
            ft.replace(R.id.frame_container, frag);
            fragmentChanges = true;
        }

        mDrawerLayout.closeDrawer(mDrawerList);

        if(fragmentChanges) {
            ft.commit();
            setTitle(item.getTitle());
            mDrawerListAdapter.selectedMenuId = item.getId(); //highlight item persistently
            mDrawerListAdapter.notifyDataSetChanged(); // force redraw
        }

        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "displayFragmentForNavDrawer() " + item.getTitle());
        }
    }

    // tests whether status is available and whether it changed since the last event.
    private void determineStatus() {
        try {
            if(mIsBound) {
                Integer newComputingStatus = monitor.getComputingStatus();
                if(!newComputingStatus.equals(clientComputingStatus)) {
                    // computing status has changed, update and invalidate to force adaption of action items
                    clientComputingStatus = newComputingStatus;
                    supportInvalidateOptionsMenu();
                }
                if(numberProjectsInNavList != monitor.getProjects().size()) {
                    numberProjectsInNavList =
                            mDrawerListAdapter.compareAndAddProjects((ArrayList<Project>) monitor.getProjects());
                }
                //setAppTitle();
            }
        }
        catch(Exception e) {
            if(Logging.ERROR) {
                Log.e(Logging.TAG, "BOINCActivity.determineStatus error: ", e);
            }
        }
    }

    public final boolean onKeyDown(final int keyCode, final KeyEvent keyEvent) {
        if(keyCode == KeyEvent.KEYCODE_MENU) {
            if(this.mDrawerLayout.isDrawerOpen(this.mDrawerList)) {
                this.mDrawerLayout.closeDrawer(this.mDrawerList);
            }
            else {
                this.mDrawerLayout.openDrawer(this.mDrawerList);
            }
            return true;
        }
        return super.onKeyDown(keyCode, keyEvent);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onCreateOptionsMenu()");
        }

        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onPrepareOptionsMenu()");
        }

        // run mode, set title and icon based on status
        MenuItem runMode = menu.findItem(R.id.run_mode);
        if(clientComputingStatus == ClientStatus.COMPUTING_STATUS_NEVER) {
            // display play button
            runMode.setTitle(R.string.menu_run_mode_enable);
            runMode.setIcon(R.drawable.playw);
        }
        else {
            // display stop button
            runMode.setTitle(R.string.menu_run_mode_disable);
            runMode.setIcon(R.drawable.pausew);
        }

        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(Logging.DEBUG) {
            Log.d(Logging.TAG, "BOINCActivity onOptionsItemSelected()");
        }

        // toggle drawer
        if(mDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }

        switch(item.getItemId()) {
            case R.id.run_mode:
                if(item.getTitle().equals(getApplication().getString(R.string.menu_run_mode_disable))) {
                    if(Logging.DEBUG) {
                        Log.d(Logging.TAG, "run mode: disable");
                    }
                    new WriteClientModeAsync().execute(BOINCDefs.RUN_MODE_NEVER);
                }
                else if(item.getTitle().equals(getApplication().getString(R.string.menu_run_mode_enable))) {
                    if(Logging.DEBUG) {
                        Log.d(Logging.TAG, "run mode: enable");
                    }
                    new WriteClientModeAsync().execute(BOINCDefs.RUN_MODE_AUTO);
                }
                else if(Logging.DEBUG) {
                    Log.d(Logging.TAG, "run mode: unrecognized command");
                }
                return true;
            case R.id.projects_add:
                startActivity(new Intent(this, SelectionListActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        // Sync the toggle state after onRestoreInstanceState has occurred.
        mDrawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        // Pass any configuration change to the drawer toggls
        mDrawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    public void setTitle(CharSequence title) {
        mTitle = title;
        getSupportActionBar().setTitle(mTitle);
    }

    private final class WriteClientModeAsync extends AsyncTask<Integer, Void, Boolean> {

        @Override
        protected Boolean doInBackground(Integer... params) {
            // setting provided mode for both, CPU computation and network.
            Boolean runMode;
            try {
                runMode = monitor.setRunMode(params[0]);
            }
            catch(RemoteException e) {
                runMode = false;
            }
            Boolean networkMode;
            try {
                networkMode = monitor.setNetworkMode(params[0]);
            }
            catch(RemoteException e) {
                networkMode = false;
            }
            return runMode && networkMode;
        }

        @Override
        protected void onPostExecute(Boolean success) {
            if(success) {
                try {
                    monitor.forceRefresh();
                }
                catch(RemoteException e) {
                    if(Logging.ERROR) {
                        Log.e(Logging.TAG, "BOINCActivity.onKeyDown.onPostExecute() error: ", e);
                    }
                }
            }
            else if(Logging.WARNING) {
                Log.w(Logging.TAG, "setting run and network mode failed");
            }
        }
    }
}

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/BOINCActivity.kt
/*
 * This file is part of BOINC.
 * http://boinc.berkeley.edu
 * Copyright (C) 2021 University of California
 *
 * BOINC is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * BOINC is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.berkeley.boinc

import android.app.Dialog
import android.app.Service
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.content.pm.PackageManager.PackageInfoFlags
import android.content.res.Configuration
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.Window
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.net.toUri
import androidx.core.os.bundleOf
import androidx.fragment.app.replace
import androidx.lifecycle.lifecycleScope
import edu.berkeley.boinc.adapter.NavDrawerListAdapter
import edu.berkeley.boinc.adapter.NavDrawerListAdapter.NavDrawerItem
import edu.berkeley.boinc.attach.AboutActivity
import edu.berkeley.boinc.attach.AttachAccountManagerActivity
import edu.berkeley.boinc.attach.SelectionListActivity
import edu.berkeley.boinc.client.ClientStatus
import edu.berkeley.boinc.client.IMonitor
import edu.berkeley.boinc.client.Monitor
import edu.berkeley.boinc.client.MonitorAsync
import edu.berkeley.boinc.databinding.MainBinding
import edu.berkeley.boinc.ui.eventlog.EventLogActivity
import edu.berkeley.boinc.utils.Logging
import edu.berkeley.boinc.utils.RUN_MODE_AUTO
import edu.berkeley.boinc.utils.RUN_MODE_NEVER
import edu.berkeley.boinc.utils.getColorCompat
import edu.berkeley.boinc.utils.writeClientModeAsync
import kotlinx.coroutines.launch

class BOINCActivity : AppCompatActivity() {
    private var clientComputingStatus = -1
    private var numberProjectsInNavList = 0

    // app title (changes with nav bar selection)
    private var mTitle: CharSequence? = null

    private lateinit var binding: MainBinding

    // nav drawer title
    private var mDrawerTitle: CharSequence? = null
    private lateinit var mDrawerToggle: ActionBarDrawerToggle
    private lateinit var mDrawerListAdapter: NavDrawerListAdapter
    private val mConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            // This is called when the connection with the service has been established, getService returns
            // the Monitor object that is needed to call functions.
            monitor = MonitorAsync(IMonitor.Stub.asInterface(service))
            mIsBound = true
            determineStatus()
        }

        override fun onServiceDisconnected(className: ComponentName) {
            // This should not happen
            monitor = null
            mIsBound = false

            Logging.logError(Logging.Category.GUI_ACTIVITY, "BOINCActivity onServiceDisconnected")
        }
    }
    private val mClientStatusChangeRec: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            Logging.logVerbose(Logging.Category.CLIENT, "BOINCActivity ClientStatusChange - onReceive()")

            determineStatus()
        }
    }
    private val ifcsc = IntentFilter("edu.berkeley.boinc.clientstatuschange")

    public override fun onCreate(savedInstanceState: Bundle?) {
        Logging.logVerbose(Logging.Category.GUI_ACTIVITY, "BOINCActivity onCreate()")

        super.onCreate(savedInstanceState)
        binding = MainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // setup navigation bar
        mDrawerTitle = title
        mTitle = mDrawerTitle
        binding.drawerList.onItemClickListener =
                OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
            // display view for selected nav drawer item
            dispatchNavBarOnClick(mDrawerListAdapter.getItem(position), false)
        }
        mDrawerListAdapter = NavDrawerListAdapter(this)
        binding.drawerList.adapter = mDrawerListAdapter
        // enabling action bar app icon and behaving it as toggle button
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setHomeButtonEnabled(true)

        mDrawerToggle = object : ActionBarDrawerToggle(this, binding.drawerLayout,
                R.string.app_name,  // nav drawer openapplicationContext - description for accessibility
                R.string.app_name // nav drawer close - description for accessibility
        ) {
            override fun onDrawerClosed(view: View) {
                supportActionBar!!.title = mTitle
                // calling onPrepareOptionsMenu() to show action bar icons
                invalidateOptionsMenu()
            }

            override fun onDrawerOpened(drawerView: View) {
                supportActionBar!!.title = mDrawerTitle
                // force redraw of all items (adapter.getView()) in order to adapt changing icons or number of tasks/notices
                mDrawerListAdapter.notifyDataSetChanged()
                // calling onPrepareOptionsMenu() to hide action bar icons
                invalidateOptionsMenu()
            }
        }
        mDrawerToggle.drawerArrowDrawable.color = getColorCompat(R.color.white)
        binding.drawerLayout.addDrawerListener(mDrawerToggle)

        // pre-select fragment
        // 1. check if explicitly requested fragment present
        // e.g. after initial project attach.
        var targetFragId = intent.getIntExtra("targetFragment", -1)

        // 2. if no explicit request, try to restore previous selection
        if (targetFragId < 0 && savedInstanceState != null) {
            targetFragId = savedInstanceState.getInt("navBarSelectionId")
        }
        val item: NavDrawerItem? = if (targetFragId < 0) {
            // if none of the above, go to default
            mDrawerListAdapter.getItem(0)
        } else {
            mDrawerListAdapter.getItemForId(targetFragId)
        }
        if (item != null) {
            dispatchNavBarOnClick(item, true)
        } else {
            Logging.logWarning(Logging.Category.GUI_ACTIVITY, "onCreate: fragment selection returned null")
        }

        //bind monitor service
        doBindService()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt("navBarSelectionId", mDrawerListAdapter.selectedMenuId)
        super.onSaveInstanceState(outState)
    }

    override fun onDestroy() {
        Logging.logVerbose(Logging.Category.GUI_ACTIVITY, "BOINCActivity onDestroy()")

        doUnbindService()
        super.onDestroy()
    }

    override fun onNewIntent(intent: Intent) {
        Logging.logVerbose(Logging.Category.GUI_ACTIVITY, "BOINCActivity onNewIntent()")

        // onNewIntent gets called if activity is brought to front via intent, but was still alive, so onCreate is not called again
        // getIntent always returns the intent activity was created of, so this method is the only hook to receive an updated intent
        // e.g. after (not initial) project attach
        super.onNewIntent(intent)
        // navigate to explicitly requested fragment (e.g. after project attach)
        val id = intent.getIntExtra("targetFragment", -1)
        Logging.logDebug(Logging.Category.GUI_ACTIVITY, "BOINCActivity onNewIntent() for target fragment: $id")

        val item: NavDrawerItem? = if (id < 0) {
            // if ID is -1, go to default
            mDrawerListAdapter.getItem(0)
        } else {
            mDrawerListAdapter.getItemForId(id)
        }
        dispatchNavBarOnClick(item, false)
    }

    override fun onResume() { // gets called by system every time activity comes to front. after onCreate upon first creation
        super.onResume()
        registerReceiver(mClientStatusChangeRec, ifcsc)
        determineStatus()
    }

    override fun onPause() { // gets called by system every time activity loses focus.
        Logging.logDebug(Logging.Category.GUI_ACTIVITY, "BOINCActivity onPause()")

        super.onPause()
        unregisterReceiver(mClientStatusChangeRec)
    }

    private fun doBindService() {
        // start service to allow setForeground later on...
        startService(Intent(this, Monitor::class.java))
        // Establish a connection with the service, onServiceConnected gets called when
        bindService(Intent(this, Monitor::class.java), mConnection, Service.BIND_AUTO_CREATE)
    }

    private fun doUnbindService() {
        if (mIsBound) {
            // Detach existing connection.
            unbindService(mConnection)
            mIsBound = false
        }
    }

    /**
     * React to selection of nav bar item
     *
     * @param item Nav bar item
     * @param init Initialize
     */
    private fun dispatchNavBarOnClick(item: NavDrawerItem?, init: Boolean) {
        // update the main content by replacing fragments
        if (item == null) {
            Logging.logWarning(Logging.Category.USER_ACTION, "dispatchNavBarOnClick returns, item null.")

            return
        }
            Logging.logVerbose(Logging.Category.USER_ACTION,
                    "dispatchNavBarOnClick for item with id: ${item.id} title: ${item.title}" +
                            " is project? ${item.isProjectItem}")

        val ft = supportFragmentManager.beginTransaction()
        var fragmentChanges = false
        if (init) {
            // if init, setup status fragment
            ft.replace<StatusFragment>(R.id.status_container)
        }
        if (!item.isProjectItem) {
            when (item.id) {
                R.string.tab_tasks -> {
                    ft.replace<TasksFragment>(R.id.frame_container)
                    fragmentChanges = true
                }
                R.string.tab_notices -> {
                    ft.replace<NoticesFragment>(R.id.frame_container)
                    fragmentChanges = true
                }
                R.string.tab_projects -> {
                    ft.replace<ProjectsFragment>(R.id.frame_container)
                    fragmentChanges = true
                }
                R.string.menu_help -> startActivity(Intent(Intent.ACTION_VIEW, "https://boinc.berkeley.edu/wiki/BOINC_Help".toUri()))
                R.string.menu_report_issue -> startActivity(Intent(Intent.ACTION_VIEW, "https://boinc.berkeley.edu/trac/wiki/ReportBugs".toUri()))
                R.string.menu_about -> startActivity(Intent(this, AboutActivity::class.java))
                R.string.menu_eventlog -> startActivity(Intent(this, EventLogActivity::class.java))
                R.string.projects_add -> startActivity(Intent(this, SelectionListActivity::class.java))
                R.string.attachproject_acctmgr_header -> startActivity(Intent(this, AttachAccountManagerActivity::class.java))
                R.string.tab_preferences -> {
                    ft.replace<SettingsFragment>(R.id.frame_container)
                    fragmentChanges = true
                }
                else -> Logging.logError(Logging.Category.USER_ACTION,
                            "dispatchNavBarOnClick() could not find corresponding fragment for" +
                            " ${item.title}")
            }
        } else {
            // ProjectDetailsFragment. Data shown based on given master URL
            ft.replace<ProjectDetailsFragment>(R.id.frame_container,
                    args = bundleOf("url" to item.projectMasterUrl))
            fragmentChanges = true
        }
        binding.drawerLayout.closeDrawer(binding.drawerList)
        if (fragmentChanges) {
            ft.commit()
            title = item.title
            mDrawerListAdapter.selectedMenuId = item.id //highlight item persistently
            mDrawerListAdapter.notifyDataSetChanged() // force redraw
        }
        Logging.logDebug(Logging.Category.USER_ACTION, "displayFragmentForNavDrawer() " + item.title)
    }

    // tests whether status is available and whether it changed since the last event.
    private fun determineStatus() {
        try {
            if (mIsBound) {
                val newComputingStatus = monitor!!.computingStatus
                if (newComputingStatus != clientComputingStatus) {
                    // computing status has changed, update and invalidate to force adaption of action items
                    clientComputingStatus = newComputingStatus
                    invalidateOptionsMenu()
                }
                if (numberProjectsInNavList != monitor!!.projects.size) {
                    numberProjectsInNavList = mDrawerListAdapter.compareAndAddProjects(monitor!!.projects)
                }
                mDrawerListAdapter.updateUseAccountManagerItem()
            }
        } catch (e: Exception) {
            Logging.logException(Logging.Category.GUI_ACTIVITY, "BOINCActivity.determineStatus error: ", e)
        }
    }

    override fun onKeyDown(keyCode: Int, keyEvent: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_MENU) {
            if (binding.drawerLayout.isDrawerOpen(binding.drawerList)) {
                binding.drawerLayout.closeDrawer(binding.drawerList)
            } else {
                binding.drawerLayout.openDrawer(binding.drawerList)
            }
            return true
        }
        return super.onKeyDown(keyCode, keyEvent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        Logging.logVerbose(Logging.Category.GUI_ACTIVITY, "BOINCActivity onCreateOptionsMenu()")

        val inflater = menuInflater
        inflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        Logging.logVerbose(Logging.Category.GUI_ACTIVITY, "BOINCActivity onPrepareOptionsMenu()")

        // run mode, set title and icon based on status
        val runMode = menu.findItem(R.id.run_mode)
        if (clientComputingStatus == ClientStatus.COMPUTING_STATUS_NEVER) {
            // display play button
            runMode.setTitle(R.string.menu_run_mode_enable)
            runMode.setIcon(R.drawable.ic_baseline_play_arrow_white)
        } else {
            // display stop button
            runMode.setTitle(R.string.menu_run_mode_disable)
            runMode.setIcon(R.drawable.ic_baseline_pause_white)
        }
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        Logging.logVerbose(Logging.Category.USER_ACTION, "BOINCActivity onOptionsItemSelected()")

        // toggle drawer
        return if (mDrawerToggle.onOptionsItemSelected(item)) {
            true
        } else when (item.itemId) {
            R.id.run_mode -> {
                when {
                    item.title == application.getString(R.string.menu_run_mode_disable) -> {
                        Logging.logDebug(Logging.Category.USER_ACTION, "run mode: disable")

                        lifecycleScope.launch {
                            writeClientMode(RUN_MODE_NEVER)
                        }
                    }
                    item.title == application.getString(R.string.menu_run_mode_enable) -> {
                        Logging.logDebug(Logging.Category.USER_ACTION, "run mode: enable")

                        lifecycleScope.launch {
                            writeClientMode(RUN_MODE_AUTO)
                        }
                    }
                    else -> {
                        Logging.logDebug(Logging.Category.USER_ACTION, "run mode: unrecognized command")
                    }
                }
                true
            }
            R.id.projects_add -> {
                startActivity(Intent(this, SelectionListActivity::class.java))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        // Sync the toggle state after onRestoreInstanceState has occurred.
        mDrawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        // Pass any configuration change to the drawer toggls
        mDrawerToggle.onConfigurationChanged(newConfig)
    }

    override fun setTitle(title: CharSequence) {
        mTitle = title
        supportActionBar!!.title = mTitle
    }

    private fun writeClientMode(mode: Int) {
        val success = writeClientModeAsync(mode)

        if (success) {
            try {
                monitor!!.forceRefresh()
            } catch (e: RemoteException) {
                    Logging.logException(Logging.Category.GUI_ACTIVITY, "BOINCActivity.writeClientMode() error: ", e)
            }
        } else {
            Logging.logWarning(Logging.Category.GUI_ACTIVITY, "BOINCActivity setting run and network mode failed")
        }
    }

    companion object {
        @JvmField
        var monitor: MonitorAsync? = null
        var mIsBound = false
    }
}
